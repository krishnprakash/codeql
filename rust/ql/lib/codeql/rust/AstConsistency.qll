/**
 * Provides classes for recognizing control flow graph inconsistencies.
 */

private import rust
private import codeql.rust.elements.internal.generated.ParentChild

private predicate multipleToStrings(Element e) { strictcount(e.toString()) > 1 }

/**
 * Holds if `e` has more than one `toString()` result.
 */
query predicate multipleToStrings(Element e, string cls, string s) {
  multipleToStrings(e) and
  cls = e.getPrimaryQlClasses() and
  s = strictconcat(e.toString(), ", ")
}

/**
 * Holds if `e` has more than one `Location`.
 */
query predicate multipleLocations(Locatable e) { strictcount(e.getLocation()) > 1 }

/**
 * Holds if `e` does not have a `Location`.
 */
query predicate noLocation(Locatable e) { not exists(e.getLocation()) }

private predicate multiplePrimaryQlClasses(Element e) {
  strictcount(string cls | cls = e.getAPrimaryQlClass() and cls != "VariableAccess") > 1
}

/**
 * Holds if `e` has more than one `getPrimaryQlClasses()` result.
 */
query predicate multiplePrimaryQlClasses(Element e, string s) {
  multiplePrimaryQlClasses(e) and
  s = strictconcat(e.getPrimaryQlClasses(), ", ")
}

private Element getParent(Element child) { child = getChildAndAccessor(result, _, _) }

private predicate multipleParents(Element child) { strictcount(getParent(child)) > 1 }

/**
 * Holds if `child` has more than one AST parent.
 */
query predicate multipleParents(Element child, string childClass, Element parent, string parentClass) {
  multipleParents(child) and
  childClass = child.getPrimaryQlClasses() and
  parent = getParent(child) and
  parentClass = parent.getPrimaryQlClasses()
}

/**
 * Gets counts of abstract syntax tree inconsistencies of each type.
 */
int getAstInconsistencyCounts(string type) {
  // total results from all the AST consistency query predicates.
  type = "Multiple toStrings" and
  result = count(Element e | multipleToStrings(e) | e)
  or
  type = "Multiple locations" and
  result = count(Element e | multipleLocations(e) | e)
  or
  type = "No location" and
  result = count(Element e | noLocation(e) | e)
  or
  type = "Multiple primary QL classes" and
  result = count(Element e | multiplePrimaryQlClasses(e) | e)
  or
  type = "Multiple parents" and
  result = count(Element e | multipleParents(e) | e)
}
