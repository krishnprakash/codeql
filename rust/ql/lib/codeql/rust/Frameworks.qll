/**
 * This file imports all models of frameworks and libraries.
 */

private import codeql.rust.frameworks.Reqwest
private import codeql.rust.frameworks.RustCrypto
private import codeql.rust.frameworks.stdlib.Env
private import codeql.rust.frameworks.Sqlx
