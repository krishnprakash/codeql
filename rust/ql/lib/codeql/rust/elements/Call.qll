/**
 * This module provides the public class `Call`.
 */

private import internal.CallImpl

final class Call = Impl::Call;
