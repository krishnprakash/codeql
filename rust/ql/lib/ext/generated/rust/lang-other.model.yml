# THIS FILE IS AN AUTO-GENERATED MODELS AS DATA FILE. DO NOT EDIT.
extensions:
  - addsTo:
      pack: codeql/rust-all
      extensible: summaryModel
    data:
      - ["lang:other", "<crate::detect::cache::Initializer>::set", "Argument[0]", "Argument[self]", "taint", "df-generated"]
      - ["lang:other", "<crate::detect::cache::Initializer>::unset", "Argument[0]", "Argument[self]", "taint", "df-generated"]
