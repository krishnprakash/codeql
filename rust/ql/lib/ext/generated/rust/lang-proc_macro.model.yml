# THIS FILE IS AN AUTO-GENERATED MODELS AS DATA FILE. DO NOT EDIT.
extensions:
  - addsTo:
      pack: codeql/rust-all
      extensible: summaryModel
    data:
      - ["lang:proc_macro", "<&[u8] as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&[u8] as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&[u8] as crate::bridge::rpc::DecodeMut>::decode", "Argument[0].Element", "Argument[0].Reference.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&[u8] as crate::bridge::rpc::DecodeMut>::decode", "Argument[0].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&crate::bridge::Marked as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Marked::value]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&crate::bridge::Marked as crate::bridge::rpc::Decode>::decode", "Argument[1].Field[proc_macro::bridge::server::HandleStore::FreeFunctions].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&crate::bridge::Marked as crate::bridge::rpc::Decode>::decode", "Argument[1].Field[proc_macro::bridge::server::HandleStore::SourceFile].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&crate::bridge::Marked as crate::bridge::rpc::Decode>::decode", "Argument[1].Field[proc_macro::bridge::server::HandleStore::TokenStream].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&mut crate::bridge::Marked as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Marked::value]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&mut crate::bridge::Marked as crate::bridge::rpc::DecodeMut>::decode", "Argument[1].Field[proc_macro::bridge::server::HandleStore::FreeFunctions].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&mut crate::bridge::Marked as crate::bridge::rpc::DecodeMut>::decode", "Argument[1].Field[proc_macro::bridge::server::HandleStore::SourceFile].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&mut crate::bridge::Marked as crate::bridge::rpc::DecodeMut>::decode", "Argument[1].Field[proc_macro::bridge::server::HandleStore::TokenStream].Element", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&str as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<&str as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<bool as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<bool as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<char as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<char as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Delimiter as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Delimiter as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::delimiter", "Argument[self].Field[0].Field[proc_macro::bridge::Group::delimiter]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::delimiter", "Argument[self].Field[proc_macro::Group(0)].Field[proc_macro::bridge::Group::delimiter]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::new", "Argument[0]", "ReturnValue.Field[proc_macro::Group(0)].Field[proc_macro::bridge::Group::delimiter]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::new", "Argument[1].Field[0]", "ReturnValue.Field[proc_macro::Group(0)].Field[proc_macro::bridge::Group::stream]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::new", "Argument[1].Field[proc_macro::TokenStream(0)]", "ReturnValue.Field[proc_macro::Group(0)].Field[proc_macro::bridge::Group::stream]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::stream", "Argument[self].Field[0].Field[proc_macro::bridge::Group::stream]", "ReturnValue.Field[proc_macro::TokenStream(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Group>::stream", "Argument[self].Field[proc_macro::Group(0)].Field[proc_macro::bridge::Group::stream]", "ReturnValue.Field[proc_macro::TokenStream(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Ident>::new", "Argument[1].Field[0]", "ReturnValue.Field[proc_macro::Ident(0)].Field[proc_macro::bridge::Ident::span]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Ident>::new", "Argument[1].Field[proc_macro::Span(0)]", "ReturnValue.Field[proc_macro::Ident(0)].Field[proc_macro::bridge::Ident::span]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Ident>::new_raw", "Argument[1].Field[0]", "ReturnValue.Field[proc_macro::Ident(0)].Field[proc_macro::bridge::Ident::span]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Ident>::new_raw", "Argument[1].Field[proc_macro::Span(0)]", "ReturnValue.Field[proc_macro::Ident(0)].Field[proc_macro::bridge::Ident::span]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Ident>::set_span", "Argument[0]", "Argument[self]", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::Ident>::span", "Argument[self].Field[0].Field[proc_macro::bridge::Ident::span]", "ReturnValue.Field[proc_macro::Span(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Ident>::span", "Argument[self].Field[proc_macro::Ident(0)].Field[proc_macro::bridge::Ident::span]", "ReturnValue.Field[proc_macro::Span(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Literal>::set_span", "Argument[0]", "Argument[self]", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::Literal>::span", "Argument[self].Field[0].Field[proc_macro::bridge::Literal::span]", "ReturnValue.Field[proc_macro::Span(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Literal>::span", "Argument[self].Field[proc_macro::Literal(0)].Field[proc_macro::bridge::Literal::span]", "ReturnValue.Field[proc_macro::Span(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Punct>::as_char", "Argument[self]", "ReturnValue", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::Punct>::new", "Argument[0]", "ReturnValue", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::Punct>::set_span", "Argument[0]", "Argument[self]", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::Punct>::span", "Argument[self].Field[0].Field[proc_macro::bridge::Punct::span]", "ReturnValue.Field[proc_macro::Span(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Punct>::span", "Argument[self].Field[proc_macro::Punct(0)].Field[proc_macro::bridge::Punct::span]", "ReturnValue.Field[proc_macro::Span(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Spacing as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::Spacing as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenStream as crate::to_tokens::ToTokens>::into_token_stream", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenStream>::expand_expr", "Argument[self].Field[0].Field[core::option::Option::Some(0)]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenStream>::expand_expr", "Argument[self].Field[proc_macro::TokenStream(0)].Field[core::option::Option::Some(0)]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenTree as crate::convert::From>::from", "Argument[0]", "ReturnValue.Field[proc_macro::TokenTree::Group(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenTree as crate::convert::From>::from", "Argument[0]", "ReturnValue.Field[proc_macro::TokenTree::Ident(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenTree as crate::convert::From>::from", "Argument[0]", "ReturnValue.Field[proc_macro::TokenTree::Literal(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::TokenTree as crate::convert::From>::from", "Argument[0]", "ReturnValue.Field[proc_macro::TokenTree::Punct(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::boxed::Box as crate::convert::From>::from", "Argument[0].Field[proc_macro::bridge::rpc::PanicMessage::StaticStr(0)]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::boxed::Box as crate::convert::From>::from", "Argument[0].Field[proc_macro::bridge::rpc::PanicMessage::String(0)]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::DelimSpan>::from_single", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::DelimSpan::close]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::DelimSpan>::from_single", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::DelimSpan::entire]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::DelimSpan>::from_single", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::DelimSpan::open]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Diagnostic as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Diagnostic::level]", "ReturnValue.Field[proc_macro::bridge::Diagnostic::level]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Diagnostic as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Diagnostic::message]", "ReturnValue.Field[proc_macro::bridge::Diagnostic::message]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Group as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Group::delimiter]", "ReturnValue.Field[proc_macro::bridge::Group::delimiter]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Ident as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Ident::is_raw]", "ReturnValue.Field[proc_macro::bridge::Ident::is_raw]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::LitKind as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::LitKind as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Literal as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Literal::kind]", "ReturnValue.Field[proc_macro::bridge::Literal::kind]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Marked as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::Marked::value]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Marked as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Marked::value]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Marked as crate::bridge::rpc::DecodeMut>::decode", "Argument[1]", "ReturnValue", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::bridge::Punct as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Punct::ch]", "ReturnValue.Field[proc_macro::bridge::Punct::ch]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::Punct as crate::bridge::Unmark>::unmark", "Argument[self].Field[proc_macro::bridge::Punct::joint]", "ReturnValue.Field[proc_macro::bridge::Punct::joint]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::Client as crate::clone::Clone>::clone", "Argument[self].Reference", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::attr", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::client::ProcMacro::Attr::name]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::bang", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::client::ProcMacro::Bang::name]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::custom_derive", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::client::ProcMacro::CustomDerive::trait_name]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::custom_derive", "Argument[1]", "ReturnValue.Field[proc_macro::bridge::client::ProcMacro::CustomDerive::attributes]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::name", "Argument[self].Field[proc_macro::bridge::client::ProcMacro::Attr::name]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::name", "Argument[self].Field[proc_macro::bridge::client::ProcMacro::Bang::name]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::ProcMacro>::name", "Argument[self].Field[proc_macro::bridge::client::ProcMacro::CustomDerive::trait_name]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::SourceFile as crate::clone::Clone>::clone", "Argument[self].Reference", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::client::TokenStream as crate::clone::Clone>::clone", "Argument[self].Reference", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::closure::Closure as crate::convert::From>::from", "Argument[0]", "ReturnValue", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::bridge::closure::Closure>::call", "Argument[0]", "Argument[self]", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::bridge::closure::Closure>::call", "Argument[self]", "ReturnValue", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::bridge::fxhash::FxHasher as crate::hash::Hasher>::finish", "Argument[self]", "ReturnValue", "taint", "df-generated"]
      - ["lang:proc_macro", "<crate::bridge::handle::InternedStore>::copy", "Argument[self].Field[proc_macro::bridge::handle::InternedStore::owned].Element", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::handle::InternedStore>::new", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::handle::InternedStore::owned].Field[proc_macro::bridge::handle::OwnedStore::counter]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::handle::OwnedStore>::new", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::handle::OwnedStore::counter]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::rpc::PanicMessage>::as_str", "Argument[self].Field[proc_macro::bridge::rpc::PanicMessage::StaticStr(0)]", "ReturnValue.Field[core::option::Option::Some(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::rpc::PanicMessage>::as_str", "Argument[self].Field[proc_macro::bridge::rpc::PanicMessage::String(0)]", "ReturnValue.Field[core::option::Option::Some(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::server::Dispatcher as crate::bridge::server::DispatcherTrait>::dispatch", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::server::MaybeCrossThread as crate::bridge::server::ExecutionStrategy>::run_bridge_and_client", "Argument[2].ReturnValue", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::server::MaybeCrossThread>::new", "Argument[0]", "ReturnValue.Field[proc_macro::bridge::server::MaybeCrossThread::cross_thread]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::bridge::server::SameThread as crate::bridge::server::ExecutionStrategy>::run_bridge_and_client", "Argument[2].ReturnValue", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Children as crate::iter::traits::iterator::Iterator>::next", "Argument[self].Field[0].Element", "ReturnValue.Field[core::option::Option::Some(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Children as crate::iter::traits::iterator::Iterator>::next", "Argument[self].Field[proc_macro::diagnostic::Children(0)].Element", "ReturnValue.Field[core::option::Option::Some(0)]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::error", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::help", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::level", "Argument[self].Field[proc_macro::diagnostic::Diagnostic::level]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::message", "Argument[self].Field[proc_macro::diagnostic::Diagnostic::message]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::new", "Argument[0]", "ReturnValue.Field[proc_macro::diagnostic::Diagnostic::level]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::note", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::set_level", "Argument[0]", "Argument[self].Field[proc_macro::diagnostic::Diagnostic::level]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::span_error", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::span_help", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::span_note", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::span_warning", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::spanned", "Argument[1]", "ReturnValue.Field[proc_macro::diagnostic::Diagnostic::level]", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::spans", "Argument[self].Field[proc_macro::diagnostic::Diagnostic::spans]", "ReturnValue.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Diagnostic>::warning", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Level as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::diagnostic::Level as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::string::String as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::string::String as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<crate::vec::Vec as crate::diagnostic::MultiSpan>::into_spans", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<u32 as crate::bridge::rpc::DecodeMut>::decode", "Argument[0].Element", "Argument[0].Reference.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<u8 as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<u8 as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<u8 as crate::bridge::rpc::DecodeMut>::decode", "Argument[0].Element", "Argument[0].Reference.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "<u8 as crate::bridge::rpc::DecodeMut>::decode", "Argument[0].Element", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<usize as crate::bridge::Mark>::mark", "Argument[0]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<usize as crate::bridge::Unmark>::unmark", "Argument[self]", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "<usize as crate::bridge::rpc::DecodeMut>::decode", "Argument[0].Element", "Argument[0].Reference.Reference", "value", "dfc-generated"]
      - ["lang:proc_macro", "crate::bridge::client::state::set", "Argument[1].ReturnValue", "ReturnValue", "value", "dfc-generated"]
      - ["lang:proc_macro", "crate::bridge::client::state::set", "Argument[1]", "Argument[1].Parameter[0]", "value", "dfc-generated"]
      - ["lang:proc_macro", "crate::bridge::client::state::with", "Argument[0].ReturnValue", "ReturnValue", "value", "dfc-generated"]
  - addsTo:
      pack: codeql/rust-all
      extensible: sinkModel
    data:
      - ["lang:proc_macro", "<crate::Ident>::new", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::Ident>::new_raw", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::Literal>::f32_suffixed", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::Literal>::f32_unsuffixed", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::Literal>::f64_suffixed", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::Literal>::f64_unsuffixed", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::Punct>::new", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<crate::bridge::symbol::Symbol>::new_ident", "Argument[0]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<f32 as crate::to_tokens::ToTokens>::to_tokens", "Argument[self]", "log-injection", "df-generated"]
      - ["lang:proc_macro", "<f64 as crate::to_tokens::ToTokens>::to_tokens", "Argument[self]", "log-injection", "df-generated"]
