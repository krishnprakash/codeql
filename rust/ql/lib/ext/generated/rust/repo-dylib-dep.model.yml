# THIS FILE IS AN AUTO-GENERATED MODELS AS DATA FILE. DO NOT EDIT.
extensions:
  - addsTo:
      pack: codeql/rust-all
      extensible: summaryModel
    data:
      - ["repo::dylib-dep", "crate::foo", "Argument[0]", "Argument[1].Parameter[0]", "value", "dfc-generated"]
