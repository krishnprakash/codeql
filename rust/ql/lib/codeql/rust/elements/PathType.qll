// generated by codegen, do not edit
/**
 * This module provides the public class `PathType`.
 */

private import internal.PathTypeImpl
import codeql.rust.elements.Path
import codeql.rust.elements.TypeRef

/**
 * A type referring to a path. For example:
 * ```rust
 * type X = std::collections::HashMap<i32, i32>;
 * type Y = X::Item;
 * ```
 */
final class PathType = Impl::PathType;
