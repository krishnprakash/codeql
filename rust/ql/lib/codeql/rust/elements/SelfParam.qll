// generated by codegen, do not edit
/**
 * This module provides the public class `SelfParam`.
 */

private import internal.SelfParamImpl
import codeql.rust.elements.Lifetime
import codeql.rust.elements.Name
import codeql.rust.elements.ParamBase

/**
 * A `self` parameter. For example `self` in:
 * ```rust
 * fn push(&mut self, value: T) {
 *   // ...
 * }
 * ```
 */
final class SelfParam = Impl::SelfParam;
