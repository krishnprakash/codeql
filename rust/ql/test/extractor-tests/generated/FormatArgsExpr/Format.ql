// generated by codegen, do not edit
import codeql.rust.elements
import TestUtils

from Format x, FormatArgsExpr getParent, int getIndex, string hasArgument
where
  toBeTested(x) and
  not x.isUnknown() and
  getParent = x.getParent() and
  getIndex = x.getIndex() and
  if x.hasArgument() then hasArgument = "yes" else hasArgument = "no"
select x, "getParent:", getParent, "getIndex:", getIndex, "hasArgument:", hasArgument
