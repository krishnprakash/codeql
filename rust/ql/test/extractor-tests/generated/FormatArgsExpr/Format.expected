| gen_format.rs:5:21:5:22 | {} | getParent: | gen_format.rs:5:14:5:32 | FormatArgsExpr | getIndex: | 1 | hasArgument: | no |
| gen_format_args_expr.rs:6:19:6:20 | {} | getParent: | gen_format_args_expr.rs:6:17:6:37 | FormatArgsExpr | getIndex: | 1 | hasArgument: | no |
| gen_format_args_expr.rs:6:26:6:29 | {:?} | getParent: | gen_format_args_expr.rs:6:17:6:37 | FormatArgsExpr | getIndex: | 3 | hasArgument: | no |
| gen_format_args_expr.rs:7:19:7:21 | {b} | getParent: | gen_format_args_expr.rs:7:17:7:43 | FormatArgsExpr | getIndex: | 1 | hasArgument: | yes |
| gen_format_args_expr.rs:7:27:7:31 | {a:?} | getParent: | gen_format_args_expr.rs:7:17:7:43 | FormatArgsExpr | getIndex: | 3 | hasArgument: | yes |
| gen_format_args_expr.rs:9:19:9:21 | {x} | getParent: | gen_format_args_expr.rs:9:17:9:28 | FormatArgsExpr | getIndex: | 1 | hasArgument: | yes |
| gen_format_args_expr.rs:9:24:9:26 | {y} | getParent: | gen_format_args_expr.rs:9:17:9:28 | FormatArgsExpr | getIndex: | 3 | hasArgument: | yes |
| gen_format_argument.rs:5:21:5:46 | {value:#width$.precision$} | getParent: | gen_format_argument.rs:5:14:5:47 | FormatArgsExpr | getIndex: | 1 | hasArgument: | yes |
| gen_format_argument.rs:7:21:7:30 | {0:#1$.2$} | getParent: | gen_format_argument.rs:7:14:7:56 | FormatArgsExpr | getIndex: | 1 | hasArgument: | yes |
